types/src/sample.rs

```diff
@@ -2,45 +2,153 @@ use std::mem::size_of;
 
 use blockstore::block::CidError;
 use bytes::{BufMut, BytesMut};
+use celestia_proto::proof::pb::Proof as RawProof;
+use celestia_proto::share::p2p::shwap::Sample as RawSample;
 use cid::CidGeneric;
 use multihash::Multihash;
+use nmt_rs::nmt_proof::NamespaceProof as NmtNamespaceProof;
+use nmt_rs::NamespaceMerkleHasher;
+use serde::{Deserialize, Serialize};
+use tendermint::Hash;
+use tendermint_proto::Protobuf;
 
-use crate::axis::{AxisId, AxisType};
-use crate::DataAvailabilityHeader;
-use crate::{Error, Result};
+use crate::extended_data_square::{AxisType, ExtendedDataSquare};
+use crate::nmt::{NamespaceProof, NamespacedHash, NamespacedHashExt, NamespacedSha2Hasher, Nmt};
+use crate::row::RowId;
+use crate::{Error, Result, Share};
 
 const SAMPLE_ID_SIZE: usize = SampleId::size();
 pub const SAMPLE_ID_MULTIHASH_CODE: u64 = 0x7801;
 pub const SAMPLE_ID_CODEC: u64 = 0x7800;
 
-/// Represents particular sample along the axis on specific Data Square
+/// Represents a location of a sample on the EDS
 #[derive(Debug, PartialEq, Clone, Copy)]
 pub struct SampleId {
-    pub axis: AxisId,
+    pub row: RowId,
     pub index: u16,
 }
 
-impl SampleId {
-    /// Create new SampleId. Index references sample number from the entire Data Square (is
-    /// converted to row/col coordinates internally). Same location can be sampled row or
-    /// column-wise, axis_type is used to distinguish that. Axis root hash is calculated from the
-    /// DataAvailabilityHeader
-    #[allow(dead_code)] // unused for now
+/// Represents Sample, with proof of its inclusion and location on EDS
+#[derive(Serialize, Deserialize, Clone)]
+#[serde(try_from = "RawSample", into = "RawSample")]
+pub struct Sample {
+    pub sample_id: SampleId,
+
+    pub sample_proof_type: AxisType,
+    pub share: Share,
+    pub proof: NamespaceProof,
+}
+
+impl Sample {
+    /// Create a new Sample. Index references sample number from the entire Data Square
     pub fn new(
         axis_type: AxisType,
         index: usize,
-        dah: &DataAvailabilityHeader,
+        eds: &ExtendedDataSquare,
         block_height: u64,
     ) -> Result<Self> {
-        let square_len = dah.square_len();
+        let square_len = eds.square_len();
 
         let (axis_index, sample_index) = match axis_type {
             AxisType::Row => (index / square_len, index % square_len),
             AxisType::Col => (index % square_len, index / square_len),
         };
 
+        let shares = eds.axis(axis_type, axis_index)?;
+
+        let mut tree = Nmt::with_hasher(NamespacedSha2Hasher::with_ignore_max_ns(true));
+
+        // XXX: are erasure coded shares correctly prefixed with parity namespace?
+        for s in &shares {
+            tree.push_leaf(s.data(), *s.namespace())
```

Comment:
the other half of shares needs to be inserted with `*Namespace::PARITY_SHARE`.
Please correct it together with a test that the proof can be verified with the root from DAH

Reply:
just saw it answers the comment above :sweat_smile: No. Parity shares are generated, so the namespace they have is just a random bytes from our point of view

Reply:
Got it thanks for clearing that up!